pkg/operation/botanist/extension.go

```diff
@@ -187,3 +199,85 @@ func (b *Botanist) WaitUntilExtensionResourcesDeleted(ctx context.Context) error
 
 	return flow.Parallel(fns...)(ctx)
 }
+
+// AnnotateExtensionCRDsForMigration annotates resources of extensions CRDs with annotation operation equal to migrate
+func (b *Botanist) AnnotateExtensionCRDsForMigration(ctx context.Context) error {
+	var fns []flow.TaskFn
+
+	for _, listObj := range []runtime.Object{
+		&extensionsv1alpha1.BackupEntryList{},
+		&extensionsv1alpha1.ControlPlaneList{},
+		&extensionsv1alpha1.ExtensionList{},
+		&extensionsv1alpha1.InfrastructureList{},
+		&extensionsv1alpha1.NetworkList{},
+		&extensionsv1alpha1.OperatingSystemConfigList{},
+		&extensionsv1alpha1.WorkerList{},
+	} {
+		listObjCopy := listObj
+		if err := b.K8sSeedClient.Client().List(ctx, listObjCopy, client.InNamespace(b.Shoot.SeedNamespace)); err != nil {
+			return err
+		}
+
+		fns = append(fns, func(ctx context.Context) error {
+			return meta.EachListItem(listObjCopy, annotateObjectForMigrationFunc(ctx, b.K8sSeedClient.Client()))
+		})
+	}
+
+	return flow.Parallel(fns...)(ctx)
+}
+
+func annotateObjectForMigrationFunc(ctx context.Context, client client.Client) func(runtime.Object) error {
+	return func(obj runtime.Object) error {
+		return kutil.TryUpdate(ctx, k8sretry.DefaultBackoff, client, obj, func() error {
+			acc, err := extensions.Accessor(obj)
+			if err != nil {
+				return err
+			}
+
+			kutil.SetMetaDataAnnotation(acc, v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationMigrate)
+			return nil
+		})
+	}
+}
+
+func (b *Botanist) restoreExtensionObject(ctx context.Context, client client.Client, obj runtime.Object, objMeta *metav1.ObjectMeta, objStatus *extensionsv1alpha1.DefaultStatus, resourceKind, resourceName string, purpose *string) error {
+	if err := b.restoreExtensionObjectState(ctx, client, obj, objStatus, resourceKind, objMeta.GetName(), purpose); err != nil {
+		return err
+	}
+
+	return b.annotateExtensionObjectWithOperationRestore(ctx, client, obj, objMeta)
```

Comment:
Hm, when the extension objects are created in the new seed the they don't have a status. I think that [this predicate](https://github.com/gardener/gardener/blob/master/extensions/pkg/predicate/predicate.go#L168-L178) will make the extension controller to immediately start the reconciliation. This would be bad as e.g. a new infrastructure would be created. Can we prevent the extensions from starting to recconcile? Maybe create the object with some new annotation "wait-for-state" that tells the extension to do nothing unless the `.status.state` was synced. Then the `gardener.cloud/operation=restore` annotation can be set to trigger the restoration process?

Reply:
While we were testing the whole scenario @vlvasilev changed this predicate to be 
`return lastOp != nil && lastOp.GetState() != gardencorev1beta1.LastOperationStateSucceeded` and it was working fine.

Reply:
Not sure if the change of the condition is a good idea in all scenarios. What if the lastOperation could not be added/updated during creation?

Reply:
See the comment above in the same context.

Reply:
@swilen-iwanow I'm not sure if this is the right way to go, looks/sounds more like a workaround. I guess it an explicit handling would be better contract-wise. WDYT?

Reply:
Just to sum up. To trigger the migration, change of the spec.seedName is required, which will set the lastOperation type to `Migrate` and trigger the "prepare flow". We annotate the ext objects with `gardener.cloud/operation=migrate`. The controllers will stop reconciling and the ShootState sync controller should stop also. Then we remove the source seed name from the `Shoot.Status.SeedName` by setting it to nil. This will increase the generationId and will trigger new reconciliation. We determine that if lastOperation was Migrate and it was Successful we start are in restoration phase. Each object is then created with annotation `wait-for-state` and controllers (also the sync controller to prevent erasing the state from the ShootState) should skip them until status.state and annotation restore are set.